File: libmmux/src/mmux/client/mmux_client.cc
#include "mmux_client.h"

namespace mmux {
namespace client {

using namespace directory;

mmux_client::mmux_client(const std::string &host, int dir_port, int lease_port)
    : fs_(std::make_shared<directory_client>(host, dir_port)),
      lease_worker_(host, lease_port) {
  lease_worker_.start();
}

std::shared_ptr<directory::directory_client> mmux_client::fs() {
  return fs_;
}

directory::lease_renewal_worker &mmux_client::lease_worker() {
  return lease_worker_;
}

void mmux_client::begin_scope(const std::string &path) {
  lease_worker_.add_path(path);
}

void mmux_client::end_scope(const std::string &path) {
  lease_worker_.remove_path(path);
}

std::shared_ptr<storage::kv_client> mmux_client::create(const std::string &path,
                                                        const std::string &backing_path,
                                                        size_t num_blocks,
                                                        size_t chain_length,
                                                        int32_t flags) {
  auto s = fs_->create(path, backing_path, num_blocks, chain_length, flags);
  begin_scope(path);
  return std::make_shared<storage::kv_client>(fs_, path, s);
}

std::shared_ptr<storage::kv_client> mmux_client::open(const std::string &path) {
  auto s = fs_->open(path);
  begin_scope(path);
  return std::make_shared<storage::kv_client>(fs_, path, s);
}

std::shared_ptr<storage::kv_client> mmux_client::open_or_create(const std::string &path,
                                                                const std::string &backing_path,
                                                                size_t num_blocks,
                                                                size_t chain_length,
                                                                int32_t flags) {
  auto s = fs_->open_or_create(path, backing_path, num_blocks, chain_length, flags);
  begin_scope(path);
  return std::make_shared<storage::kv_client>(fs_, path, s);
}

std::shared_ptr<storage::kv_listener> mmux_client::listen(const std::string &path) {
  auto s = fs_->open(path);
  begin_scope(path);
  return std::make_shared<storage::kv_listener>(path, s);
}

void mmux_client::remove(const std::string &path) {
  end_scope(path);
  fs_->remove(path);
}

void mmux_client::sync(const std::string &path, const std::string &dest) {
  fs_->sync(path, dest);
}

void mmux_client::dump(const std::string &path, const std::string &dest) {
  fs_->dump(path, dest);
}

void mmux_client::load(const std::string &path, const std::string &dest) {
  fs_->load(path, dest);
}

void mmux_client::close(const std::string &path) {
  end_scope(path);
}

}
}

File: jmmux/core/src/main/java/mmux/MMuxClient.java
package mmux;

import com.github.phantomthief.thrift.client.ThriftClient;
import com.github.phantomthief.thrift.client.impl.ThriftClientImpl;
import com.github.phantomthief.thrift.client.pool.ThriftConnectionPoolProvider;
import com.github.phantomthief.thrift.client.pool.ThriftServerInfo;
import com.github.phantomthief.thrift.client.pool.impl.DefaultThriftConnectionPoolImpl;
import java.io.Closeable;
import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import java.util.function.Supplier;
import mmux.directory.directory_service;
import mmux.directory.rpc_data_status;
import mmux.kv.KVClient;
import mmux.lease.LeaseWorker;
import mmux.notification.KVListener;
import org.apache.commons.pool2.impl.GenericKeyedObjectPoolConfig;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.transport.TFramedTransport;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

public class MMuxClient implements Closeable {

  private ThriftClient fs;
  private LeaseWorker worker;
  private Thread workerThread;

  public MMuxClient(String host, int dirPort, int leasePort) throws TException {
    fs = makeFS(host, dirPort);
    worker = new LeaseWorker(host, leasePort);
    workerThread = new Thread(worker);
    workerThread.start();
  }

  private ThriftClient makeFS(String host, int port) {
    Function<ThriftServerInfo, TTransport> tProvider = info -> {
      TSocket socket = new TSocket(info.getHost(), info.getPort());
      return new TFramedTransport(socket);
    };
    GenericKeyedObjectPoolConfig pConf = new GenericKeyedObjectPoolConfig();
    ThriftServerInfo info = ThriftServerInfo.of(host, port);
    Supplier<List<ThriftServerInfo>> iProvider = () -> Collections.singletonList(info);
    ThriftConnectionPoolProvider pProvider = new DefaultThriftConnectionPoolImpl(pConf, tProvider);
    return new ThriftClientImpl(iProvider, pProvider);
  }

  public directory_service.Iface fs() {
    return fs.iface(directory_service.Client.class, TBinaryProtocol::new, 0);
  }

  LeaseWorker getWorker() {
    return worker;
  }

  public void beginScope(String path) {
    worker.addPath(path);
  }

  public void endScope(String path) {
    worker.removePath(path);
  }

  public KVClient create(String path) throws TException {
    return create(path, "", 1, 1, 0);
  }

  public KVClient create(String path, String backingPath) throws TException {
    return create(path, backingPath, 1, 1, 0);
  }

  public KVClient create(String path, String backingPath, int numBlocks, int chainLength)
      throws TException {
    return create(path, backingPath, numBlocks, chainLength, 0);
  }

  public KVClient create(String path, String backingPath, int numBlocks, int chainLength, int flags)
      throws TException {
    rpc_data_status status = fs().create(path, backingPath, numBlocks, chainLength, flags);
    beginScope(path);
    return new KVClient(fs, path, status, 1000);
  }

  public KVClient open(String path) throws TException {
    rpc_data_status status = fs().open(path);
    beginScope(path);
    return new KVClient(fs, path, status, 1000);
  }

  public KVClient openOrCreate(String path) throws TException {
    return openOrCreate(path, "", 1, 1, 0);
  }

  public KVClient openOrCreate(String path, String backingPath) throws TException {
    return openOrCreate(path, backingPath, 1, 1, 0);
  }

  public KVClient openOrCreate(String path, String backingPath, int numBlocks, int chainLength)
      throws TException {
    return openOrCreate(path, backingPath, numBlocks, chainLength, 0);
  }

  public KVClient openOrCreate(String path, String backingPath, int numBlocks,
      int chainLength, int flags) throws TException {
    rpc_data_status status = fs().openOrCreate(path, backingPath, numBlocks, chainLength, flags);
    beginScope(path);
    return new KVClient(fs, path, status, 1000);
  }

  public KVListener listen(String path) throws TException {
    rpc_data_status status = fs().open(path);
    beginScope(path);
    return new KVListener(path, status);
  }

  public void remove(String path) throws TException {
    endScope(path);
    fs().remove(path);
  }

  public void removeAll(String path) throws TException {
    worker.removePaths(path);
    fs().removeAll(path);
  }

  public void rename(String oldPath, String newPath) throws TException {
    fs().rename(oldPath, newPath);
    worker.renamePath(oldPath, newPath);
  }

  public void sync(String path, String backingPath) throws TException {
    fs().sync(path, backingPath);
  }

  public void dump(String path, String backingPath) throws TException {
    fs().dump(path, backingPath);
  }

  public void load(String path, String backingPath) throws TException {
    fs().load(path, backingPath);
  }

  public void close(String path) {
    endScope(path);
  }

  @Override
  public void close() throws IOException {
    worker.stop();
    try {
      workerThread.join();
    } catch (InterruptedException e) {
      throw new IOException(e);
    }
  }
}
